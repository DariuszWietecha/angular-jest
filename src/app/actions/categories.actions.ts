import { createAction, props } from '@ngrx/store';
import { ICategory } from '../services/categories.service';

export const loadCategories = createAction(
  '[Categories] Load Categories'
);
export const loadCategoriesSuccess = createAction(
  '[Categories] Load Categories Success',
  props<{ data: ICategory[] }>()
);
export const loadCategoriesFailure = createAction(
  '[Categories] Load Categories Failure',
  props<{ error: any }>()
);

export const loadCategory = createAction(
  '[Categories] Load Category',
  props<{ categoryId?: string }>()
);
export const loadCategorySuccess = createAction(
  '[Categories] Load Category Success',
  props<ICategory>()
);
export const loadCategoryFailure = createAction(
  '[Categories] Load Category Failure',
  props<{ error: any }>()
);

export const createCategory = createAction(
  '[Categories] Create Category',
  props<{ data: ICategory }>()
);
export const createCategorySuccess = createAction(
  '[Categories] Create Category Success',
);
export const createCategoryFailure = createAction(
  '[Categories] Create Category Failure',
  props<{ error: any }>()
);

export const editCategory = createAction(
  '[Categories] Edit Category',
  props<{ data: ICategory }>()
);
export const editCategorySuccess = createAction(
  '[Categories] Edit Category Success'
);
export const editCategoryFailure = createAction(
  '[Categories] Edit Category Failure',
  props<{ error: any }>()
);
