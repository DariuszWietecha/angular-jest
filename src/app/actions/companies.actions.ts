import { createAction, props } from '@ngrx/store';
import { ICompany } from '../services/companies.service';

export const loadCompanies = createAction(
  '[Companies] Load Companiess'
);
export const loadCompaniesSuccess = createAction(
  '[Companies] Load Companiess Success',
  props<{ data: ICompany[] }>()
);
export const loadCompaniesFailure = createAction(
  '[Companies] Load Companiess Failure',
  props<{ error: any }>()
);

export const loadCompany = createAction(
  '[Companies] Load Company',
  props<{ companyId?: string }>()
);
export const loadCompanySuccess = createAction(
  '[Companies] Load Company Success',
  props<ICompany>()
);
export const loadCompanyFailure = createAction(
  '[Companies] Load Company Failure',
  props<{ error: any }>()
);

export const createCompany = createAction(
  '[Companies] Create Company',
  props<{ data: ICompany }>()
);
export const createCompanySuccess = createAction(
  '[Companies] Create Company Success'
);
export const createCompanyFailure = createAction(
  '[Companies] Create Company Failure',
  props<{ error: any }>()
);

export const editCompany = createAction(
  '[Companies] Edit Company',
  props<{ data: ICompany }>()
);
export const editCompanySuccess = createAction(
  '[Companies] Edit Company Success'
);
export const editCompanyFailure = createAction(
  '[Companies] Edit Company Failure',
  props<{ error: any }>()
);

