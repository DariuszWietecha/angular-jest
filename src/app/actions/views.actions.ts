import { createAction, props } from '@ngrx/store';
import { IHomeData } from '../services/views.service';

export const loadHomeData = createAction(
  '[Home] Load Home data',
  props<{ categoryId?: string }>()
);

export const loadHomeDataSuccess = createAction(
  '[Views] Load Home data Success',
  props<IHomeData>()
);

export const loadHomeDataFailure = createAction(
  '[Views] Load Home data Failure',
  props<{ error: any }>()
);
