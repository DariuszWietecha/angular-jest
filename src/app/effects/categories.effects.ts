import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { CategoriesService } from '../services/categories.service';
import * as categoriesActions from '../actions/categories.actions';
import { map, mergeMap, catchError } from 'rxjs/operators';
import { of } from 'rxjs';

@Injectable()
export class CategoriesEffects {
  @Effect()
  loadCategories$ = this.actions$
    .pipe(
      ofType(categoriesActions.loadCategories),
      mergeMap(() => this.categoriesService.list()
        .pipe(
          map((data) => categoriesActions.loadCategoriesSuccess({data})),
          catchError((error) => of(categoriesActions.loadCategoriesFailure({ error })))
        ))
    );

  @Effect()
  loadCategory$ = this.actions$
    .pipe(
      ofType(categoriesActions.loadCategory),
      mergeMap((action) => this.categoriesService.get(action.categoryId)
        .pipe(
          map((data) => categoriesActions.loadCategorySuccess(data)),
          catchError((error) => of(categoriesActions.loadCategoryFailure({ error })))
        ))
    );

  @Effect()
  createCategory$ = this.actions$
    .pipe(
      ofType(categoriesActions.createCategory),
      mergeMap((action) => this.categoriesService.create(action.data)
        .pipe(
          map(() => categoriesActions.createCategorySuccess()),
          catchError((error) => of(categoriesActions.createCategoryFailure({ error })))
        ))
    );

  @Effect()
  editCategory$ = this.actions$
    .pipe(
      ofType(categoriesActions.editCategory),
      mergeMap((action) => this.categoriesService.edit(action.data)
        .pipe(
          map(() => categoriesActions.editCategorySuccess()),
          catchError((error) => of(categoriesActions.editCategoryFailure({ error })))
        ))
    );

  constructor(
    private actions$: Actions,
    private categoriesService: CategoriesService
  ) { }
}
