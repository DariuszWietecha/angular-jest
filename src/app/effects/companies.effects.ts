import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { CompaniesService } from '../services/companies.service';
import * as companiesActions from '../actions/companies.actions';
import { map, mergeMap, catchError } from 'rxjs/operators';
import { of } from 'rxjs';


@Injectable()
export class CompaniesEffects {
  @Effect()
  loadCompanies$ = this.actions$
    .pipe(
      ofType(companiesActions.loadCompanies),
      mergeMap(() => this.companiesService.list()
        .pipe(
          map((data) => companiesActions.loadCompaniesSuccess({ data })),
          catchError((error) => of(companiesActions.loadCompaniesFailure({ error })))
        ))
    );

  @Effect()
  loadCompany$ = this.actions$
    .pipe(
      ofType(companiesActions.loadCompany),
      mergeMap((action) => this.companiesService.get(action.companyId)
        .pipe(
          map((data) => companiesActions.loadCompanySuccess(data)),
          catchError((error) => of(companiesActions.loadCompanyFailure({ error })))
        ))
    );

  @Effect()
  createCompany$ = this.actions$
    .pipe(
      ofType(companiesActions.createCompany),
      mergeMap((action) => this.companiesService.create(action.data)
        .pipe(
          map(() => companiesActions.createCompanySuccess()),
          catchError((error) => of(companiesActions.createCompanyFailure({ error })))
        ))
    );

  @Effect()
  editCompany$ = this.actions$
    .pipe(
      ofType(companiesActions.editCompany),
      mergeMap((action) => this.companiesService.edit(action.data)
        .pipe(
          map(() => companiesActions.editCompanySuccess()),
          catchError((error) => of(companiesActions.editCompanyFailure({ error })))
        ))
    );

  constructor(
    private actions$: Actions,
    private companiesService: CompaniesService
  ) { }
}
