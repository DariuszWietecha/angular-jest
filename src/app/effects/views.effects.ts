import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { ViewsService } from '../services/views.service';
import { loadHomeData, loadHomeDataSuccess, loadHomeDataFailure } from '../actions/views.actions';
import { map, mergeMap, catchError } from 'rxjs/operators';
import { of } from 'rxjs';

@Injectable()
export class ViewsEffects {
  @Effect()
  loadHomeData$ = this.actions$
    .pipe(
      ofType(loadHomeData),
      mergeMap((action) => this.viewsService.get(action.categoryId)
        .pipe(
          map((data) => loadHomeDataSuccess(data)),
          catchError((error) => of(loadHomeDataFailure({ error })))
        ))
    );

  constructor(
    private actions$: Actions,
    private viewsService: ViewsService
  ) { }
}
