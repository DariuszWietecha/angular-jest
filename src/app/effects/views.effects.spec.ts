import { TestBed } from '@angular/core/testing';
import { cold, hot } from 'jasmine-marbles';
import { Observable } from 'rxjs';
import { HttpClientModule } from '@angular/common/http';
import { Actions } from '@ngrx/effects';
import { ViewsEffects } from './views.effects';
import { ViewsService } from '../services/views.service';
import { empty } from 'rxjs';
import * as viewsActions from '../actions/views.actions';

export class TestActions extends Actions {
  constructor() {
    super(empty());
  }

  set stream(source: Observable<any>) {
    this.source = source;
  }
}

export function getActions() {
  return new TestActions();
}

describe('ViewsEffects', () => {
  let actions$: TestActions;
  let effects: ViewsEffects;
  let viewsService: ViewsService;
  const homeDataMock = {
    categories: require('../../../test/data/categories/categories-get.json'),
    companies: require('../../../test/data/companies/companies-get.json'),
    categoryId: 'a0308830-660b-11ea-9266-adf5decb4537',
  };

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [HttpClientModule],
      providers: [
        ViewsEffects,
        {
          provide: Actions,
          useFactory: getActions
        },
        {
          provide: ViewsService,
          useValue: { get: jest.fn() },
        },
      ]
    });
  });

  beforeEach(() => {
    effects = TestBed.get(ViewsEffects);
    viewsService = TestBed.get(ViewsService);
    actions$ = TestBed.get(Actions);
  });

  it('should be created', () => {
    expect(effects).toBeTruthy();
  });

  it.only('loadHomeDataSuccess', () => {
    const action = viewsActions.loadHomeData({ categoryId: '123' });
    const completion = viewsActions.loadHomeDataSuccess(homeDataMock);
    const response = cold('a|', { a: homeDataMock });
    const expected = cold('-b', { b: completion });

    actions$.stream = hot('-a', { a: action });

    viewsService.get = jest.fn(() => response);

    // expect(effects.loadHomeData$).toMatchSnapshot();
    expect(effects.loadHomeData$).toEqual(expected);
  });

  it('loadHomeDataFailure', () => {
    const action = viewsActions.loadHomeData({ categoryId: '123' });
    const completion = viewsActions.loadHomeDataFailure({ error: 'Error' });
    const response = cold('#');
    const expected = cold('-b', { b: completion });

    actions$.stream = hot('-a', { a: action });

    viewsService.get = jest.fn(() => response);

    expect(effects.loadHomeData$).toMatchSnapshot();
    // expect(effects.loadHomeData$).toEqual(expected);
  });
});
