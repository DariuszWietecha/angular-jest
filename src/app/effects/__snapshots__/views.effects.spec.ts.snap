// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewsEffects loadHomeDataFailure 1`] = `
Actions {
  "_isScalar": false,
  "operator": MergeMapOperator {
    "concurrent": Infinity,
    "project": [Function],
  },
  "source": Actions {
    "_isScalar": false,
    "operator": FilterOperator {
      "predicate": [Function],
      "thisArg": undefined,
    },
    "source": TestActions {
      "_isScalar": false,
      "source": TestHotObservable {
        "_isScalar": false,
        "error": undefined,
        "marbles": "-a",
        "source": SubscriptionLoggable {
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "messages": Array [
            Object {
              "frame": 10,
              "notification": Notification {
                "error": undefined,
                "hasValue": true,
                "kind": "N",
                "value": Object {
                  "categoryId": "123",
                  "type": "[Home] Load Home data",
                },
              },
            },
          ],
          "observers": Array [],
          "scheduler": TestScheduler {
            "SchedulerAction": [Function],
            "actions": Array [],
            "active": false,
            "assertDeepEqual": [Function],
            "coldObservables": Array [
              SubscriptionLoggable {
                "_isScalar": false,
                "_subscribe": [Function],
                "messages": Array [
                  Object {
                    "frame": 0,
                    "notification": Notification {
                      "error": "error",
                      "hasValue": false,
                      "kind": "E",
                      "value": undefined,
                    },
                  },
                ],
                "scheduler": [Circular],
                "subscriptions": Array [],
              },
              SubscriptionLoggable {
                "_isScalar": false,
                "_subscribe": [Function],
                "messages": Array [
                  Object {
                    "frame": 10,
                    "notification": Notification {
                      "error": undefined,
                      "hasValue": true,
                      "kind": "N",
                      "value": Object {
                        "error": "Error",
                        "type": "[Views] Load Home data Failure",
                      },
                    },
                  },
                ],
                "scheduler": [Circular],
                "subscriptions": Array [],
              },
            ],
            "flushTests": Array [],
            "frame": 0,
            "hotObservables": Array [
              [Circular],
            ],
            "index": -1,
            "maxFrames": 750,
            "now": [Function],
            "runMode": false,
            "scheduled": undefined,
          },
          "subscriptions": Array [],
          "thrownError": null,
        },
        "values": Object {
          "a": Object {
            "categoryId": "123",
            "type": "[Home] Load Home data",
          },
        },
      },
    },
  },
}
`;
